sdk/trace/experimental_span_test.go

```diff
@@ -0,0 +1,53 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"go.opentelemetry.io/otel/trace"
+)
+
+func TestAddLinks(t *testing.T) {
+	ctx, tp := context.Background(), NewTracerProvider()
+	defer func(ctx context.Context, tp *TracerProvider) {
+		if err := tp.Shutdown(ctx); err != nil {
+			panic(err)
+		}
+	}(ctx, tp)
+
+	ctx, span := tp.Tracer("test").Start(ctx, "test_add_links")
+	defer span.End()
+
+	links := []trace.Link{
+		trace.LinkFromContext(ctx),
+	}
+
+	if s, ok := span.(trace.ExperimentalSpan); ok {
+		s.AddLinks(links...)
+	}
```

Comment:
Something like below would not require making any changes in the API.

```suggestion
	if s, ok := span.(interface {
		AddLinks(links ...Link)
	}); ok {
		s.AddLinks(links...)
	}
```

Reply:
It will still, however, require the `AddLinks` method to be exported. Something that would immediately need to be stable given the state of the SDK version.

We cannot accept a method that we know will conflict with the eventual method needed to be added to support the API when it is extended to include an `AddLinks` method.

Reply:
`recordingSpan` type is not exported so it will not be a ABI change.

Reply:
I considered that, and think it could be a viable alternative to a formal interface, but the tradeoff is how easy it is for users who have opted in to realize we've dropped/changed support.

With type-checking for `ExperimentalSpan`, users would get a compilation error, whereas type-checking for `interface { AddLinks(links ...Link) }`, it could go unnoticed that we've changed or removed this signature altogether.

But, if the project policies won't allow for this sort of non-stable interface in a stable module, as @MrAlias alluded to in https://github.com/open-telemetry/opentelemetry-go/pull/4889#issuecomment-1930864364, then perhaps it would just be an added requirement for users trying to opt into these surfaces to log their own warnings for if/when they go missing:

```diff
	if s, ok := span.(interface {
		AddLinks(links ...Link)
	}); ok {
		s.AddLinks(links...)
-	}
+	} else {
+		// user is responsible for logging a warning that the type-check failed 
+	}
```

Reply:
So someone who is using a `v1` version of a Go package and does what you propose above in there code then upgrades that package to the next minor or patch version but has their code fail to compile because that method had changed or been removed wouldn't have a valid argument to make that we violated our versioning and stability guarantees(!)?

Reply:
It will compile.

Reply:
For clarification, with anonymous interface in type assertion it will continue to compile.

Reply:
> So someone who is using a `v1` version of a Go package and does what you propose above in their code then upgrades that package to the next minor or patch version but has their code fail to compile because that method had changed or been removed wouldn't have a valid argument to make that we violated our versioning and stability guarantees(!)?

Well, that's kind of the entire idea I was trying to navigate with this PR: _can this project ship opt-in features where part of the contract for opting in is knowing that the opt-in features **are not** subject to the same stability guarantees._

Fully respect that you disagree _with_ the objective, but this^ _is_ explicitly what this PR was proposing, providing a mechanism for users to opt-into features that might break their build next time they upgrade.

> I'm sorry, but no. That is not acceptable and we can talk more about it in the next SIG meeting if it needs clarification.

Sure, though, I think you might be coming at this just a little hot, given that this PR (should've hit the draft button, I suppose) was just meant for discussion. Was in no way expecting this PR to actually land in `main` as-is, or necessarily at all, was just looking to discuss some ideas.

_Something that would be useful as a happy-and-grateful community member would be for some SIG clarification on why some language SDKs are more or less willing to support experimental features (e.g. we've been able to `AddLinks()` from the dotnet SDK for a while now)._